### ISSUE TITLE
Type `Floats` cannot support Unitful vectors

### ISSUE THREAD

**Issue by JeffFessler** [emmt/LazyAlgebra.jl#4]
The `Floats` type used here is a union of AbstractFloats and Complex and is a very sensible type for the goals of this package.  Unfortunately that union is not quite general enough to include Unitful vectors because such vector's elements are subtypes of `Number` not of `AbstractFloat`.  So simple methods like `vscale!` do not apply as currently written.  MWE:

```julia
using LazyAlgebra
using Unitful: m
x = ones(5) * 1m
vscale!(x, 2)

ERROR: MethodError: no method matching vscale!(::Vector{Quantity{Float64, 𝐋, Unitful.FreeUnits{(m,), 𝐋, nothing}}}, ::Int64, ::Vector{Quantity{Float64, 𝐋, Unitful.FreeUnits{(m,), 𝐋, nothing}}})
```

It probably would be a major undertaking to support Unitful vectors here, so I am not really advocating for that.
But I am just reporting the issue here because it is relevant to another PR elsewhere:

https://github.com/emmt/LinearInterpolators.jl/pull/9

**Comment by emmt** [comment]
You are right, it is only in optimization code that complexes should be considered as pairs of floats. The inner product (`vdot` in `LazyAlgebra`) should be implemented differently in `LazyAlgebra` which is more designed for direct modeling and, say, in [`OptimPackNextGen`](https://github.com/emmt/OptimPackNextGen.jl) which is dedicated to numerical optimization of large scale problems.

I really like your duck-typing approach, type assertions should be relaxed to exploit one of the power of Julia that the same code can serve to many different argument types (I was probably too much influenced by my experience in C coding). I agree that it makes sense to let Julia complain when argument types are incompatible.

Below is a demonstration that relaxing types to any `Number` (including `Unitful` ones) as you suggested is easy doable:

```julia
function vdot1(x::AbstractArray{Tx,N}, y::AbstractArray{Ty,N}) where {Tx,Ty,N}
    s = zero(promote_type(Tx, Ty))
    @inbounds @simd for i in eachindex(x, y); s += x[i]*y[i]; end
    return s
end

function vdot2(x::AbstractArray{Tx,N}, y::AbstractArray{Ty,N}) where {Tx,Ty,N}
    s = zero(typeof(oneunit(Tx)*oneunit(Ty))) # there are probably simpler expressions that yield this
    @inbounds @simd for i in eachindex(x, y); s += x[i]*y[i]; end
    return s
end

using Unitful
T, n = Float32, 5_000
x=rand(T,n).*1u"kg"
y=rand(T,n).*1u"m"
vdot1(x,y) # throws "ERROR: ArgumentError: zero(Quantity{Float32, D, U} where {D, U}) not defined."
vdot2(x,y) # yields a correct result in "kg⋅m"
```

The only thing to adapt is the initiallization of the result (here the accumulation variable `s`) to the correct type.  The `promote_type` method fails here but the replacement is easy to write.  Besides, the `vdot2` code does not even know the `Unitful` package which avoids the dependency issue that you mentioned.

I benchmarked that the `@simd` durective was really effective here: 306ns for 5,000 entris that a pretty nice 32.7 Gflops. This means that my "type enforcing" policy is not needed to execute highly optimized code (I also checked that `vdot2` also worked for non-standard abstract arrays like `view(x,1:3:n)` without complaining).

The only remaining issue I can anticipate is the case of multipliers. To not penalize optimizations, these factors are supposed to not enforce their precision but instead follow that of the quantities multipled by them. We have to design a new implementation of `promote_multipler` that works with an extended range of scalar types. Using `Require`, the method can be automatically extended for, say, `Unitful` types when this package is loaded.  So this could be done without depending on `Unitful` or equivalent (just `Require`). Checks that multipliers have special values (0, ±1, etc) have to be modified, but I guess that this should be as simple as the replacement for `promote_type` above.  For example `α == 0` and  `α == 1` should be replaced by `α == zero(α)` and `α == oneunit(α)`. Other impacted methods are probably `vcreate` and `output_type`.  But that may not a big deal...
